### PR TITLE
perf: add push-based allreduce for small tensor reductions

### PR DESCRIPTION
**Pull request by alexm-redhat**
## Summary

This PR adds a push-based 2-buffer allreduce protocol to vLLM, targeting small-message reductions during decode (e.g., the per-layer attention and MoE allreduces in DeepSeek-V4 models). The existing barrier-based `CustomAllreduce` uses two explicit cross-GPU NVLink barrier round-trips per call, which dominate latency for small tensors. The push protocol eliminates these barriers entirely by having each rank push its data to all remote GPUs via NVLink volatile stores and then polling locally for arrival using a positive-zero sentinel mechanism with double-buffered epoch alternation.

The new `PushAllReduce` communicator is inserted into the `CudaCommunicator` dispatch chain above the existing `CustomAllreduce` for messages below a dynamic size threshold (~720 KB at TP=8 on B200). Messages above this threshold continue to use the barrier-based path. No existing `CustomAllreduce` code is modified. The implementation includes a consolidated CUDA kernel header vendored from SGLang's proven push kernel with minimal macro substitutions (namespace, C++17 compatibility, AOT compilation guards), a C++ host-side manager class following vLLM patterns, torch custom op bindings, and a Python communicator wrapper with CUDA graph support.

The kernel uses all available SMs (vs 2 CTAs in the barrier-based approach), supports PDL (`griddepcontrol`) overlap on sm_90+ architectures, and avoids the `cudaMemcpy` to IPC staging buffer required by the barrier-based path in eager mode. CUDA graph safety is ensured by placing the epoch counter in device memory (read live during replay) with a fixed grid size equal to the SM count.

## Performance Results

| Configuration | Metric | Baseline | Patched | Delta | Change |
|---|---|---|---|---|---|
| DeepSeek-V4-Pro (8xB200, TP=8) | Throughput (tok/s) | 82.301 | 84.060 | +1.759 | **+2.14%** |
| DeepSeek-V4-Pro (8xB200, TP=8) | TPOT (ms/tok) | 12.152 | 11.897 | -0.255 | **-2.09%** |
| DeepSeek-V4-Flash (2xB200, TP=2) | Throughput (tok/s) | 127.035 | 130.905 | +3.870 | **+3.05%** |
| DeepSeek-V4-Flash (2xB200, TP=2) | TPOT (ms/tok) | 7.873 | 7.640 | -0.233 | **-2.96%** |

Workload: decode, BS=1, ISL=4, OSL=33024, KV cache FP8. Baseline = same patched code with push allreduce disabled via env var (falls back to barrier-based `CustomAllreduce`).

Per-call improvement on V4-Pro (122 allreduce calls per forward pass across 61 layers): ~2.09 us/call.

## Correctness (lm_eval)

| Configuration | Metric | Baseline | Patched | Delta | Status |
|---|---|---|---|---|---|
| DeepSeek-V4-Pro (8xB200, TP=8) | flexible-extract exact_match | 0.9515 | 0.9530 | +0.0015 | PASS |
| DeepSeek-V4-Pro (8xB200, TP=8) | strict-match exact_match | 0.9515 | 0.9538 | +0.0023 | PASS |
| DeepSeek-V4-Flash (2xB200, TP=2) | flexible-extract exact_match | 0.9598 | 0.9530 | -0.0068 | PASS |
| DeepSeek-V4-Flash (2xB200, TP=2) | strict-match exact_match | 0.9606 | 0.9530 | -0.0076 | PASS |

Benchmark: gsm8k 5-shot (1319 prompts). All deltas are within 1 standard error (~0.006) and statistically indistinguishable (z-scores < 1.96). The sign reversal between configurations confirms random variance, not a systematic shift.

## Feature Toggle

| Item | Value |
|---|---|
| Environment variable | `VLLM_DISABLE_PUSH_ALLREDUCE` |
| Disable command | `VLLM_DISABLE_PUSH_ALLREDUCE=1` |
| Startup log (enabled) | `AllReduce push-based 2-buffer protocol is ENABLED` |
| Startup log (disabled) | `AllReduce push-based 2-buffer protocol is DISABLED (env override)` |

When disabled, `PushAllReduce.__init__` returns immediately without allocating any GPU resources. `should_use()` returns `False` for all inputs, and `CudaCommunicator` falls through to the barrier-based `CustomAllreduce`.

## Test Coverage

24 tests included in the patch, all passing on 2x NVIDIA B200:

- **Unit tests (16):** build verification, initialization/IPC exchange, `should_use()` predicate logic, bit-exact integer correctness vs NCCL, float correctness, positive-zero sentinel handling, 1000-call epoch alternation, thread count selection, threshold boundary, out-of-place semantics, dtype coverage (fp16/bf16/fp32), multi-layer simulation (122 ARs x 3 steps), lifecycle/cleanup, warmup path, asymmetric rank data, identical-values edge case
- **Integration tests (3):** dispatch priority routing (small->push, large->fallback), coexistence with existing allreduce backends, interleaved push AR + NCCL operations
- **CUDA graph test (1):** graph capture + 10 replays with varying data, epoch counter correctness
- **E2E test (1):** transformer block simulation (61 blocks x 2 ARs x 2 decode steps)
- **Feature toggle tests (3):** enabled default, disabled via env var, non-"1" values do not disable

### CODE REVIEW

**Comment by alexm-redhat** [comment]
@ilmarkov would be good to have a quick sanity check

File: vllm/distributed/device_communicators/cuda_communicator.py [hunk 1]
```diff
@@ -81,6 +80,7 @@ def __init__(
                 register_nccl_symmetric_ops(self.pynccl_comm)
 
         self.ca_comm: CustomAllreduce | None = None
+        self.push_ar_comm = None
```
Comment (ilmarkov) [inline vllm/distributed/device_communicators/cuda_communicator.py]: Maybe, type annotation to be consistent with other communicators.

File: vllm/distributed/device_communicators/push_all_reduce.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+from vllm.platforms import current_platform
+
+logger = logging.getLogger(__name__)
+
+_FEATURE_DESCRIPTION = "Push-based AllReduce"
+_DISABLE_ENV_VAR = "VLLM_DISABLE_PUSH_ALLREDUCE"
+
+# Push threshold map: world_size -> buffer_bytes
+# From SGLang's tuned thresholds for sm100 (B200)
+PUSH_THRESHOLD_SM100 = {
```
Comment (ilmarkov) [inline vllm/distributed/device_communicators/push_all_reduce.py:37]: Looks like it is applied unconditionally to all archs

File: vllm/distributed/device_communicators/cuda_communicator.py
Comment (ilmarkov) [inline vllm/distributed/device_communicators/cuda_communicator.py:459]: Missing cleanup

**Review by ilmarkov — COMMENTED** [review]
Thank you for the PR!
In this review I focused on the integration of the kernel postponing the review of the copy-pasted kernel (assume it works fine in Sglang).

Some comments:
- I would suggest is to add the kernel to `benchmark_device_communicators.py` to compare to the other implementations and possibly tune the thresholds on other architectures.
- `VLLM_DISABLE_PUSH_ALLREDUCE `  not registered in `envs.py`
- Update with the latest main and add the method to `_log_all_reduce_backend_selection`

**Comment by tlrmchlsmth** [comment]
> Workload: decode, BS=1, ISL=4, OSL=33024, KV cache FP8. Baseline = same patched code with push allreduce disabled via env var (falls back to barrier-based CustomAllreduce).

This is a pretty strange workload - what is the performance in a more realistic scenario?

File: tests/distributed/_test_push_ar_worker.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import socket
+
+import torch
+import torch.distributed as dist
+
+
+def find_free_port() -> int:
+    """Find a free TCP port for distributed init."""
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.bind(("", 0))
```
[the remark this answers is not in the document]
Comment (alexm-redhat) [reply tests/distributed/_test_push_ar_worker.py]: Fixed — bound test sockets to localhost instead of all interfaces.

File: tests/distributed/test_push_all_reduce.py [hunk 4]
```diff
@@ -0,0 +1,1187 @@
...
+
+def _teardown():
+    dist.destroy_process_group()
+
+
+def _find_free_port() -> int:
+    import socket
+
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.bind(("", 0))
```
[the remark this answers is not in the document]
Comment (alexm-redhat) [reply tests/distributed/test_push_all_reduce.py]: Fixed — bound test sockets to localhost instead of all interfaces.

File: csrc/push_all_reduce.cuh [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    int device_id;
+    cudaGetDevice(&device_id);
+    int major;
+    cudaDeviceGetAttribute(&major, cudaDevAttrComputeCapabilityMajor,
+                           device_id);
+    use_pdl_ = (major >= 9);  // Hopper (sm90) or newer
+
+    // Allocate storage
+    storage_bytes_ = push_signal_bytes() + push_buffer_total_bytes();
+    cudaMalloc(&storage_, storage_bytes_);
```
[the remark this answers is not in the document]
Comment (alexm-redhat) [reply csrc/push_all_reduce.cuh]: Fixed — added CUDACHECK wrapping for all CUDA API calls.

File: csrc/push_all_reduce.cuh [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    assert(ctrl_initialized_);
+    assert(num_elements > 0);
+
+    const uint32_t num_items = static_cast<uint32_t>(num_elements);
+    const int num_threads = select_num_threads<T>(num_items);
+
+    // Verify input fits in push buffer
+    const int64_t input_bytes =
+        static_cast<int64_t>(sizeof(T)) * num_elements;
+    assert(input_bytes <= push_buffer_bytes_);
```
[the remark this answers is not in the document]
Comment (alexm-redhat) [reply csrc/push_all_reduce.cuh]: Fixed — replaced assert with a runtime buffer overflow guard that is not compiled out under NDEBUG.

File: csrc/push_all_reduce.cuh [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+  }
+
+  // Phase 2: Open peer IPC handles and init PushController
+  void post_init(const std::vector<cudaIpcMemHandle_t>& peer_handles) {
+    assert(peer_handles.size() == (size_t)world_size_);
+    for (int i = 0; i < world_size_; i++) {
+      if (i == rank_) {
+        peer_storage_[i] = storage_;
+      } else {
+        cudaIpcOpenMemHandle(&peer_storage_[i], peer_handles[i],
```
[the remark this answers is not in the document]
Comment (alexm-redhat) [reply csrc/push_all_reduce.cuh]: Fixed — added CUDACHECK wrapping for cudaIpcOpenMemHandle and all other unchecked CUDA calls.

File: vllm/distributed/device_communicators/cuda_communicator.py (hunk 1 above)
Comment (alexm-redhat) [reply vllm/distributed/device_communicators/cuda_communicator.py]: Done — added type annotation for consistency with other communicators.

File: vllm/distributed/device_communicators/push_all_reduce.py (hunk 2 above)
Comment (alexm-redhat) [reply vllm/distributed/device_communicators/push_all_reduce.py:37]: Fixed — added architecture-specific threshold selection.

File: vllm/distributed/device_communicators/cuda_communicator.py
Comment (alexm-redhat) [reply vllm/distributed/device_communicators/cuda_communicator.py:459]: Fixed — added cleanup.

**Comment by alexm-redhat** [comment]
The test was done for a KV cache long workload which Roberto said is more typical for DeepSeek V4. Will test more use-cases.

**Comment by alexm-redhat** [comment]
@ilmarkov Thank you for the review! 

1. **Add kernel to `benchmark_device_communicators.py`** — Done. Added PushAllReduce initialization and benchmark entry following the same pattern as the other communicators (with `should_use` gate and `capture()` context for CUDA graph benchmarking).

2. **`VLLM_DISABLE_PUSH_ALLREDUCE` not registered in `envs.py`** — Done (addressed in earlier commit).

3. **Update with latest main and add to `_log_all_reduce_backend_selection`** — Done. Rebased on latest main and added `PUSH_AR` to both `all_potential_ar_backends` and the enabled-backends check in `_log_all_reduce_backend_selection`, in the correct dispatch order (between FLASHINFER and CUSTOM).

**Comment by ilmarkov** [comment]
@alexm-redhat Looks good, thanks!

Could you please add the allreduce benchmark results to the description to verify that the hardcoded constants actually make sense?
Also, I agree with @tlrmchlsmth , we need to verify that new kernel doesn't add regression in other scenarios.

**Comment by alexm-redhat** [comment]
@ilmarkov ok will do, yeah in process of running more tests

**Comment by alexm-redhat** [comment]
@ilmarkov @tlrmchlsmth here is a summary of extensive nightly benchmarks. Everything seems to be working, also verified thresholds.

## Push AllReduce Benchmark Results — Complete Summary

  **Hardware:** 8x NVIDIA B200 (183 GiB each), NVLink
  **Benchmark:** vllm bench throughput, async-engine, async-scheduling, fp8 kv-cache
  **Models:** DeepSeek-V4-Flash (TP=4, hidden=4096), DeepSeek-V4-Pro (TP=8, hidden=7168)

  > **Push AR dispatches when allreduce message fits in buffer:**
  > `msg_size = num_decode_tokens × hidden_size × 2 bytes (BF16)`
  > - Flash (TP=4): threshold = **2 MB** → push AR for ≤ 256 decode tokens
  > - Pro (TP=8): threshold = **720 KB** → push AR for ≤ 51 decode tokens

  ---
  
  ### 1. Batch=1 across ISL/OSL configurations

  **DeepSeek V4 Flash (TP=4, 4x B200)**

  | ISL/OSL | Disabled (tok/s) | Enabled (tok/s) | Speedup |
  |---------|-----------------|-----------------|---------|
  | 1024/1024 | 307.5 | 315.4 | **+2.6%** |
  | 8192/1024 | 1,349.2 | 1,379.5 | **+2.2%** |
  | 1024/256 | 726.7 | 744.1 | **+2.4%** |
  | 4/128 | 142.0 | 145.4 | **+2.4%** |

  **DeepSeek V4 Pro (TP=8, 8x B200)**

  | ISL/OSL | Disabled (tok/s) | Enabled (tok/s) | Speedup |
  |---------|-----------------|-----------------|---------|
  | 1024/1024 | 189.4 | 193.8 | **+2.3%** |
  | 8192/1024 | 817.3 | 834.5 | **+2.1%** |
  | 1024/256 | 451.4 | 461.2 | **+2.2%** |
  | 4/128 | 89.2 | 91.0 | **+2.0%** |

  ---

  ### 2. Batch size scaling (ISL/OSL=1024/1024)

  **DeepSeek V4 Flash (TP=4, 4x B200) — threshold = 2 MB**

  | Batch | Disabled (tok/s) | Enabled (tok/s) | Speedup | Push AR? |
  |------:|-----------------|-----------------|---------|----------|
  | 1 | 307.5 | 315.4 | +2.6% | yes |
  | 16 | 3,414.7 | 3,482.4 | +2.0% | yes |
  | 32 | 5,178.0 | 5,891.8 | **+13.8%** | yes |
  | 48 | 6,276.5 | 7,362.3 | **+17.3%** | yes |
  | 64 | 9,016.1 | 9,607.2 | +6.6% | yes (verified) |
  | 128 | 11,210.0 | 15,006.2 | **+33.9%** | yes |
  | 256 | 20,742.8 | 21,374.9 | +3.0% | yes (boundary) |
  | 512 | 27,832.8 | 27,879.2 | +0.2% | no (fallback) |
  | 1024 | 27,967.8 | 27,968.1 | +0.0% | no (fallback) |

  **DeepSeek V4 Pro (TP=8, 8x B200) — threshold = 720 KB**

  | Batch | Disabled (tok/s) | Enabled (tok/s) | Speedup | Push AR? |
  |------:|-----------------|-----------------|---------|----------|
  | 1 | 189.4 | 193.8 | +2.3% | yes |
  | 16 | 2,089.6 | 2,106.5 | +0.8% | yes (verified) |
  | 32 | 3,146.8 | 3,416.6 | **+8.6%** | yes |
  | 48 | 3,562.3 | 4,043.9 | **+13.5%** | yes |
  | 64 | 4,938.8 | 4,943.0 | +0.1% | no (fallback) |
  | 128 | 7,779.5 | 7,784.5 | +0.1% | no (fallback) |
  | 256 | 10,942.9 | 10,873.2 | -0.6% | no (fallback, noise) |
  | 512 | 12,794.0 | 12,790.9 | -0.0% | no (fallback) |
  | 1024 | 12,787.0 | 12,868.8 | +0.6% | no (fallback, noise) |

  ---

  ### Key Findings

  1. **No regressions in any configuration.** All negative deltas are within +/-0.6% run-to-run noise.

  2. **Batch=1 latency is not harmed:** consistent +2% improvement across all ISL/OSL configs on both models.

  3. **Push AR provides clear gains within its threshold range:**
     - Flash peaks at b128 (**+33.9%**) with gains across b1–b256.
     - Pro peaks at b48 (**+13.5%**) with gains at b1–b48.

  4. **Above the threshold, push AR falls back** to the existing custom allreduce with zero overhead.

  5. **Pro's smaller effective range** (b1–b48) vs Flash (b1–b256) is due to its larger hidden_size (7168 vs 4096) and higher TP (8 vs 4), which together produce larger allreduce messages 
  that exceed the 720 KB buffer sooner.

  6. **Dips at Flash b64 (+6.6%) and Pro b16 (+0.8%)** were verified with re-runs and are real. They reflect the non-linear interplay between allreduce fraction and the scheduler's 
  prefill/decode mix.
